### ISSUE TITLE
API: nvim_get_context not parsable with vim.mpack.unpack.

### ISSUE THREAD

**Issue by muniter** [neovim/neovim#15711]
### Neovim version (nvim -v)

NVIM v0.6.0-dev+276-g685cf3981

### Vim (not Nvim) behaves the same?

doesn't apply

### Operating system/version

Linux, Arch

### Terminal name/version

Alacritty

### $TERM environment variable

xterm-256color

### Installation

build from repo

### How to reproduce the issue

```
nvim -u NONE
```
```lua
vim.mpack.unpack(vim.api.nvim_get_context({types = {"jumps"}}).jumps[1])
```
Produces
```
E5108: Error executing lua [string ":lua"]:1: trailing data in msgpack string
```

I'm not familiar with mpack, but I would imagine this would work.

### Expected behavior

Be able to parse with the new msgpack functions.

### Actual behavior

```
E5108: Error executing lua [string ":lua"]:1: trailing data in msgpack string
```

**Comment by justinmk** [comment]
hmm, even this reports...

```
:echo msgpackparse([nvim_get_context({'types': ["jumps"]}).jumps[1]])
E475: Invalid argument: Incomplete msgpack string                    
[8, 1631914397, 52]
```

are you using nvim_get_context() successfully in some other manner? It is still kinda experimental (we may need to make that clearer...)

**Comment by muniter** [comment]
Was just trying to change the following test for the new feature of adding insert enter/leave to the jumplist. Checking number of jumps pre and post.

 https://github.com/neovim/neovim/blob/de406f651c6f926db4223b69ef312777e8afe01b/test/functional/terminal/ex_terminal_spec.lua#L94

**Comment by muniter** [comment]
This is the workaround:
```lua
lua vim.fn.msgpackparse(vim.api.nvim_get_context({types = {"jumps"}}).jumps)
```

**Comment by seandewar** [comment]
I don't think this is a bug, if I'm understanding this correctly?

`jumps` is a `readfile`-style list, so I don't think it's a good idea to parse only a single element, otherwise you'll be using an incomplete chunk of the dumped msgpack data.

In a `readfile`-style list, NULs in the original string are converted to NLs, and the string is split where the original NLs were. This structure is used for binary data as Vimscript strings don't support NULs.

This means the list is split in places where the dumped data happens to contain a `0x0a` byte, not where each jumplist entry is (I'm guessing this was presumed?). You'll want to parse the entire list instead.

**Comment by muniter** [comment]
@seandewar `nvim_get_context` returns a lua table with a `jumps` field with only one element, which is a msgpack string. So I'm not accessing one jump specifically but rather trying to parse the whole contents with `vim.mpack.unpack` which in this case is `jumps[1]`.

**Comment by seandewar** [comment]
Does it work if you pass that to `vim.fn.msgpackparse()` instead? I'm not sure if `vim.mpack.unpack` supports a `readfile`-style list.

**Comment by muniter** [comment]
@seandewar yes it works, I also don't get this error @justinmk  gets when just parsing "jumps"

> hmm, even this reports...
> 
> ```
> :echo msgpackparse([nvim_get_context({'types': ["jumps"]}).jumps[1]])
> E475: Invalid argument: Incomplete msgpack string                    
> [8, 1631914397, 52]
> ```
> 

Then is probably not suported by `vim.mpack.unpack`

**Comment by seandewar** [comment]
Yeah, presumably Justin had a readfile list with `len` > 1, so it was incomplete when used like that. :+1:
